File: frontend/src/api/avatar.ts
import { http } from './http'

export type AvatarGender = 'female' | 'male' | 'unspecified'

export interface VoiceOption {
  provider: TtsProvider
  value: string
  label: string
}

export type TtsProvider = 'volcengine' | 'dashscope'

export interface TtsProviderSetting {
  provider: TtsProvider
  display_name: string
  is_enabled: boolean
  is_default: boolean
  is_fallback: boolean
  model: string
  default_voice: string
  first_chunk_timeout_ms: number
  configured: boolean
}

export interface AvatarVariant {
  id: number
  digital_human_id: number
  outfit_name: string
  version: string
  original_filename: string
  file_size: number
  thumbnail_url: string | null
  validation_status: 'ready' | 'failed'
  created_at: string
}

export interface DigitalHuman {
  id: number
  name: string
  gender: AvatarGender
  role_title: string
  introduction: string | null
  tts_provider: TtsProvider
  tts_voice: string
  tts_instructions: string | null
  is_enabled: boolean
  variants: AvatarVariant[]
}

export interface ScenicAvatar {
  config_id: number
  scenic_area_id: number
  id: number
  digital_human_id: number
  name: string
  gender: AvatarGender
  role_title: string
  introduction: string | null
  outfit_name: string
  version: string
  thumbnail_url: string | null
  file_size: number
  is_enabled: boolean
  is_default: boolean
  sort_order: number
}

export interface ScenicAvatarList {
  scenic_area_id: number
  default_variant_id: number | null
  avatars: ScenicAvatar[]
}

export interface DigitalHumanPayload {
  name: string
  gender: AvatarGender
  role_title: string
  introduction?: string | null
  tts_provider: TtsProvider
  tts_voice: string
  tts_instructions?: string | null
}

const apiBaseUrl = import.meta.env.VITE_API_BASE_URL || '/api'

export function avatarAssetUrl(scenicAreaCode: string, avatarVariantId: number) {
  return `${apiBaseUrl}/avatars/scenic-areas/${encodeURIComponent(scenicAreaCode)}/variants/${avatarVariantId}/asset`
}

export function ttsProviderTestUrl(provider: TtsProvider) {
  return `${apiBaseUrl}/admin/tts/providers/${provider}/test`
}

export const avatarApi = {
  listVoices: (provider: TtsProvider = 'volcengine') => http.get<VoiceOption[]>('/admin/avatars/voices', { params: { provider } }),
  listTtsProviders: () => http.get<TtsProviderSetting[]>('/admin/tts/providers'),
  updateTtsProvider: (provider: TtsProvider, data: Partial<Omit<TtsProviderSetting, 'provider' | 'display_name' | 'configured'>>) =>
    http.patch<TtsProviderSetting>(`/admin/tts/providers/${provider}`, data),
  listHumans: () => http.get<DigitalHuman[]>('/admin/avatars/humans'),
  createHuman: (data: DigitalHumanPayload) => http.post<DigitalHuman>('/admin/avatars/humans', data),
  updateHuman: (id: number, data: Partial<DigitalHumanPayload> & { is_enabled?: boolean }) =>
    http.patch<DigitalHuman>(`/admin/avatars/humans/${id}`, data),
  deleteHuman: (id: number) => http.delete<void>(`/admin/avatars/humans/${id}`),
  listScenicConfigs: (scenicAreaId: number) => http.get<ScenicAvatar[]>('/admin/avatars/scenic-configs', { params: { scenic_area_id: scenicAreaId } }),
  uploadVariant: (data: {
    digitalHumanId: number
    scenicAreaId: number
    outfitName: string
    version: string
    thumbnailUrl?: string
    isEnabled: boolean
    isDefault: boolean
    sortOrder: number
    file: File
  }) => {
    const form = new FormData()
    form.append('digital_human_id', String(data.digitalHumanId))
    form.append('scenic_area_id', String(data.scenicAreaId))
    form.append('outfit_name', data.outfitName)
    form.append('version', data.version)
    if (data.thumbnailUrl) form.append('thumbnail_url', data.thumbnailUrl)
    form.append('is_enabled', String(data.isEnabled))
    form.append('is_default', String(data.isDefault))
    form.append('sort_order', String(data.sortOrder))
    form.append('file', data.file)
    return http.post<ScenicAvatar>('/admin/avatars/variants', form)
  },
  updateVariant: (id: number, data: { outfit_name?: string; version?: string; thumbnail_url?: string | null }) =>
    http.patch<AvatarVariant>(`/admin/avatars/variants/${id}`, data),
  deleteVariant: (id: number) => http.delete<void>(`/admin/avatars/variants/${id}`),
  updateScenicConfig: (id: number, data: { is_enabled?: boolean; is_default?: boolean; sort_order?: number }) =>
    http.patch<ScenicAvatar>(`/admin/avatars/scenic-configs/${id}`, data),
  listPublicScenicAvatars: (scenicAreaCode: string) => http.get<ScenicAvatarList>(`/avatars/scenic-areas/${scenicAreaCode}`),
}

File: backend/app/schemas/avatar.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class DigitalHumanCreate(BaseModel):
    name: str = Field(min_length=1, max_length=80)
    gender: str = Field(default="unspecified", pattern=r"^(female|male|unspecified)$")
    role_title: str = Field(min_length=1, max_length=120)
    introduction: str | None = Field(default=None, max_length=2000)
    tts_provider: str = Field(default="volcengine", pattern=r"^(volcengine|dashscope)$")
    tts_voice: str = Field(min_length=1, max_length=100)
    tts_instructions: str | None = Field(default=None, max_length=1000)


class DigitalHumanUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=80)
    gender: str | None = Field(default=None, pattern=r"^(female|male|unspecified)$")
    role_title: str | None = Field(default=None, min_length=1, max_length=120)
    introduction: str | None = Field(default=None, max_length=2000)
    tts_provider: str | None = Field(default=None, pattern=r"^(volcengine|dashscope)$")
    tts_voice: str | None = Field(default=None, min_length=1, max_length=100)
    tts_instructions: str | None = Field(default=None, max_length=1000)
    is_enabled: bool | None = None


class AvatarVariantUpdate(BaseModel):
    outfit_name: str | None = Field(default=None, min_length=1, max_length=120)
    version: str | None = Field(default=None, min_length=1, max_length=40)
    thumbnail_url: str | None = Field(default=None, max_length=500)


class ScenicAvatarConfigUpdate(BaseModel):
    is_enabled: bool | None = None
    is_default: bool | None = None
    sort_order: int | None = Field(default=None, ge=-1000, le=1000)


class VoiceOptionOut(BaseModel):
    provider: str
    value: str
    label: str


class AvatarVariantOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    digital_human_id: int
    outfit_name: str
    version: str
    original_filename: str
    file_size: int
    thumbnail_url: str | None
    validation_status: str
    created_at: datetime


class DigitalHumanOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    name: str
    gender: str
    role_title: str
    introduction: str | None
    tts_provider: str
    tts_voice: str
    tts_instructions: str | None
    is_enabled: bool
    variants: list[AvatarVariantOut] = Field(default_factory=list)


class ScenicAvatarOut(BaseModel):
    config_id: int
    scenic_area_id: int
    id: int
    digital_human_id: int
    name: str
    gender: str
    role_title: str
    introduction: str | None
    outfit_name: str
    version: str
    thumbnail_url: str | None
    file_size: int
    is_enabled: bool
    is_default: bool
    sort_order: int


class ScenicAvatarListOut(BaseModel):
    scenic_area_id: int
    default_variant_id: int | None
    avatars: list[ScenicAvatarOut]


class TtsProviderSettingUpdate(BaseModel):
    is_enabled: bool | None = None
    is_default: bool | None = None
    is_fallback: bool | None = None
    model: str | None = Field(default=None, min_length=1, max_length=120)
    default_voice: str | None = Field(default=None, min_length=1, max_length=100)
    first_chunk_timeout_ms: int | None = Field(default=None, ge=500, le=10000)


class TtsProviderSettingOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    provider: str
    display_name: str
    is_enabled: bool
    is_default: bool
    is_fallback: bool
    model: str
    default_voice: str
    first_chunk_timeout_ms: int
    last_visitor_first_chunk_ms: int | None
    last_visitor_first_chunk_at: datetime | None
    configured: bool


class TtsProviderTestRequest(BaseModel):
    text: str = Field(default="欢迎来到灵山胜境，我是您的数字讲解员。", min_length=1, max_length=200)
    voice: str | None = Field(default=None, min_length=1, max_length=100)
